integration/kv_test.go

```diff
@@ -20,18 +22,95 @@ import (
 	"github.com/cortexproject/cortex/pkg/ring/kv/etcd"
 )
 
-func TestKV_List_Delete(t *testing.T) {
-	s, err := e2e.NewScenario(networkName)
-	require.NoError(t, err)
-	defer s.Close()
+func TestKVList(t *testing.T) {
+	testKVs(t, func(t *testing.T, client kv.Client, reg *prometheus.Registry) {
+		// Create keys to list back
+		keysToCreate := []string{"key-a", "key-b", "key-c"}
+		for _, key := range keysToCreate {
+			err := client.CAS(context.Background(), key, func(in interface{}) (out interface{}, retry bool, err error) {
+				return key, false, nil
+			})
+			require.NoError(t, err, "could not create key")
+		}
 
-	// Start dependencies
-	etcdSvc := e2edb.NewETCD()
-	consulSvc := e2edb.NewConsul()
+		// Get list of keys and sort them
+		keys, err := client.List(context.Background(), "")
+		require.NoError(t, err, "could not list keys")
+		sort.Strings(keys)
+		require.Equal(t, keysToCreate, keys, "returned key paths did not match created paths")
 
-	require.NoError(t, s.StartAndWaitReady(etcdSvc, consulSvc))
+		verifyClientMetrics(t, reg, map[string]uint64{
+			"List": 1,
+			"CAS":  3,
+		})
+	})
+}
 
-	reg := prometheus.NewRegistry()
+func TestKVDelete(t *testing.T) {
+	testKVs(t, func(t *testing.T, client kv.Client, reg *prometheus.Registry) {
+		// Create a key
+		err := client.CAS(context.Background(), "key-to-delete", func(in interface{}) (out interface{}, retry bool, err error) {
+			return "key-to-delete", false, nil
+		})
+		require.NoError(t, err, "object could not be created")
+
+		// Now delete it
+		err = client.Delete(context.Background(), "key-to-delete")
+		require.NoError(t, err)
+
+		// Get it back
+		v, err := client.Get(context.Background(), "key-to-delete")
+		require.NoError(t, err, "unexpected error")
+		require.Nil(t, v, "object was not deleted")
+
+		verifyClientMetrics(t, reg, map[string]uint64{
+			"Delete": 1,
+			"CAS":    1,
+			"GET":    1,
+		})
+	})
+}
+
+func TestKVWatchAndDelete(t *testing.T) {
+	testKVs(t, func(t *testing.T, client kv.Client, reg *prometheus.Registry) {
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		err := client.CAS(context.Background(), "key-before-watch", func(in interface{}) (out interface{}, retry bool, err error) {
+			return "value-before-watch", false, nil
+		})
+		require.NoError(t, err)
+
+		w := &watcher{}
+		wg := &sync.WaitGroup{}
+		wg.Add(1)
+		go w.watch(ctx, wg, client)
+
+		err = client.CAS(context.Background(), "key-to-delete", func(in interface{}) (out interface{}, retry bool, err error) {
+			return "value-to-delete", false, nil
+		})
+		require.NoError(t, err, "object could not be created")
+
+		// Now delete it
+		err = client.Delete(context.Background(), "key-to-delete")
```

Comment:
Should we add a sleep between creating the `key-to-delete` and deleting it? I'm wondering what happens if the two operations happens very close each other: the watcher may not get any notification (depending on how the backend is implemented).

Reply:
OK, let's do that.